File: leetcode/prob500.cpp

#include <iostream>
#include <algorithm>
#include <string>
#include <vector>
#include <map>
#include <utility>
#include <iterator>
#include <unordered_map>

#include <cctype>

using namespace std;
/*
* Name: 500. Keyboard Row
* WebSite: https://leetcode.com/problems/keyboard-row/
* Description: 
*
*/
class Solution {
public:
    /*
    * Runtime:    ms
    * Your runtime beats  % of cpp submissions. 
    */
    vector<string> findWords(vector<string>& words) {
        vector<string> alphabet = {"qwertyuiop", "asdfghjkl", "zxcvbnm"};
        unordered_map<char, int> alphaToRows;
        for (size_t i = 0; i < alphabet.size(); ++i) {
            const string &alpha = alphabet[i];
            for (const auto v : alpha) {
                alphaToRows[v] = static_cast<int>(i);
            }
        }

        bool isOnlyOneRows = true; 
        vector<string> result;
        for (const auto &word : words) {
            if (!word.empty()) {
                isOnlyOneRows = true;
                for (size_t i = 0; i < word.size(); ++i) {
                    if (alphaToRows[std::tolower(word[0])] != alphaToRows[std::tolower(word[i])]) {
                        isOnlyOneRows = false;
                        break;
                    }
                }
                if (isOnlyOneRows)
                    result.emplace_back(word);
                // char first = word[0];
                // if (std::all_of(word.begin(), word.end(), [&alphaToRows, first](char w) {
                //     return alphaToRows[std::tolower(first)] == alphaToRows[std::tolower(w)];
                // })) {
                //     result.emplace_back(word);
                // } 
            }
        }
        return result;
    }
};

int main() 
{
    Solution s; 


    std::cout << "method 1: \n";
    vector<string> test = {"Hello", "Alaska", "Dad", "Peace"};
    auto ret = s.findWords(test);
    std::cout << std::endl;
    return 0;
}                

File: leetCode_Python/prob500.py
#!/usr/bin/env python3
#-*- coding:utf-8 -*-

# author    : lambda RDG
# Date&Time : 2019/05/17 16:26:37

"模块的文档注释"
# Name: 500. Keyboard Row
# Website: https://leetcode.com/problems/keyboard-row/
# Description:
import sys

class Solution:
    def findWords(self, words: List[str]) -> List[str]:
        result = []
        if not words: 
            return result
        alphabets = ("qwertyuiop", "asdfghjkl", "zxcvbnm")
        alphabetToRows = {}
        for i, alpha in enumerate(alphabets):
            for v in alpha:
                alphabetToRows[v] = i
        for word in words:
            if word:
                isOneRow = True
                for ch in word:
                    if alphabetToRows[word[0]] != alphabetToRows[ch]:
                        isOneRow = False
                        break
            if isOneRow:
                result.append(word)
        return result
        # keyboard = [set('qwertyuiop'),set('asdfghjkl'),set('zxcvbnm')]
        # output=list()
        # for word in words:
        #     if sum([1 for row in keyboard if set(word.lower()).intersection(row)])==1:
        #         output.append(word)
                
        # return output



if __name__ == "__main__":
    pass